tests/test_bootstrap.py

```diff
@@ -0,0 +1,1390 @@
+            for L_factor, M_factor, N_factor, rho_exponent in zip(
+                L_factors, M_factors, N_factors, rho_exponents
+            ):
+                results.append(
+                    test(grid_type, kwargs, L_factor, M_factor, N_factor, rho_exponent)
+                )
+
+        np.testing.assert_allclose(np.array(results), 0.5, rtol=2e-4)
+
+    @pytest.mark.unit
+    def test_BootstrapRedlConsistency_resolution(self):
+        """Confirm that the objective function is ~independent of grid resolution."""
+        helicity = (1, 0)
+        filename = ".//tests//inputs//DSHAPE_output_saved_without_current.h5"
+        eq = desc.io.load(filename)[-1]
+        eq.electron_density = PowerSeriesProfile(
+            2.0e19 * np.array([1, -0.85]), modes=[0, 4]
+        )
+        eq.electron_temperature = PowerSeriesProfile(
+            1e3 * np.array([1.02, -3, 3, -1]), modes=[0, 2, 4, 6]
+        )
+        eq.ion_temperature = PowerSeriesProfile(
+            1.1e3 * np.array([1.02, -3, 3, -1]), modes=[0, 2, 4, 6]
+        )
+        eq.atomic_number = 1.4
+
+        # Set to True to plot the profiles of <J*B>
+        if False:
+            grid = LinearGrid(L=eq.L, M=eq.M, N=0, axis=False)
+            data = eq.compute(
+                ["<J*B>", "<J*B> Redl", "rho"],
+                grid=grid,
+                helicity=helicity,
+            )
+            import matplotlib.pyplot as plt
+
+            plt.plot(data["rho"], data["<J*B>"], label="MHD")
+            plt.plot(data["rho"], data["<J*B> Redl"], label="Redl")
+            plt.xlabel("rho")
+            plt.title("<J*B>")
+            plt.legend(loc=0)
+            plt.show()
+
+        def test(grid_type, kwargs, L, M, N):
+            grid = grid_type(
+                L=L,
+                M=M,
+                N=N,
+                NFP=eq.NFP,
+                **kwargs,
+            )
+            obj = ObjectiveFunction(
+                BootstrapRedlConsistency(
+                    grid=grid,
+                    helicity=helicity,
+                    rho_exponent=1,
+                ),
+                eq,
+            )
+            scalar_objective = obj.compute_scalar(obj.x(eq))
+            return scalar_objective
+
+        results = []
+
+        # Loop over grid trypes. For LinearGrid we need to drop the
+        # point at rho=0 to avoid a divide-by-0
+        grid_types = [LinearGrid, QuadratureGrid]
+        kwargss = [{"axis": False}, {}]
+
+        # Loop over grid resolutions:
+        Ls = [10, 20, 10, 10]
+        Ms = [10, 10, 20, 10]
+        Ns = [0, 0, 0, 2]
+
+        for grid_type, kwargs in zip(grid_types, kwargss):
+            for L, M, N in zip(Ls, Ms, Ns):
+                results.append(test(grid_type, kwargs, L, M, N))
+
+        results = np.array(results)
+        np.testing.assert_allclose(results, np.mean(results), rtol=0.02)
+
+    # MJL: Should this test be decorated with @pytest.mark.slow ?
+    @pytest.mark.unit
```

Comment:
this should probably be marked as a `regression` test. We try to limit the number of full solve/optimizations we have to do for unit tests.

Reply:
This has been changed to `regression`.